src/data/gaussian_int.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Chris Hughes
+
+The gaussian integers ℤ[i].
+-/
+
+import number_theory.pell data.complex.basic algebra.euclidean_domain algebra.associated
+open zsqrtd complex
+
+@[reducible] def gaussian_int : Type := zsqrtd (-1)
+
+local notation `ℤ[i]` := gaussian_int
+
+namespace gaussian_int
+
+instance : has_repr ℤ[i] := ⟨λ x, "⟨" ++ repr x.re ++ ", " ++ repr x.im ++ "⟩"⟩
+
+instance : comm_ring ℤ[i] := zsqrtd.comm_ring
+
+def norm (x : ℤ[i]) : ℕ := x.re.nat_abs * x.re.nat_abs + x.im.nat_abs * x.im.nat_abs
+
+def to_complex (x : ℤ[i]) : ℂ := x.re + x.im * I
+
+instance : has_coe (ℤ[i]) ℂ := ⟨to_complex⟩
+
+lemma to_complex_def (x : ℤ[i]) : (x : ℂ) = x.re + x.im * I := rfl
+
+lemma to_complex_def' (x y : ℤ) : ((⟨x, y⟩ : ℤ[i]) : ℂ) = x + y * I := by simp [to_complex_def]
+
+lemma to_complex_def₂ (x : ℤ[i]) : (x : ℂ) = ⟨x.re, x.im⟩ :=
+by apply complex.ext; simp [to_complex_def]
+
+instance to_complex.is_ring_hom : is_ring_hom to_complex:=
+by refine_struct {..}; intros; apply complex.ext; simp [to_complex]
+
+instance : is_ring_hom (coe : ℤ[i] → ℂ) := to_complex.is_ring_hom
+
+@[simp] lemma int_cast_re (x : ℤ[i]) : ((x.re : ℤ) : ℝ) = (x : ℂ).re := by simp [to_complex_def]
```

Comment:
The rationale was that the coercion from ℤ to ℝ is called `int.cast`. I guess `to_real` makes sense here because otherwise it's not obvious that it's really a real specific lemma